### DISCUSSION TITLE
Transformer word embedding matrix creation

### DISCUSSION THREAD

**Discussion by urs17** [explosion/spaCy#8963] [Help: Coding & Implementations]
Hello,
I am doing NER classification with pre-trained transformer embeddings and my dataset is BIO formatted. I need to pass the embeddings to Bi-LSTM. If I use a large model, the code works well, but I am getting issues when I use the transformer model. The code below is for the large model but I need help with the transformer model.

**Large model:**

```
nlp = spacy.load('en_core_web_lg')
EMB_DIM = nlp.vocab.vectors_length
MAX_LEN = 50

def preprocess(samples):
    nlp.max_length = len(str(samples))
    tag_index = {tag: index for index, tag in enumerate(tags)}
    X = np.zeros((len(samples), MAX_LEN, EMB_DIM), dtype=np.float32)
    y = np.zeros((len(samples), MAX_LEN), dtype=np.uint8)
    vocab = nlp.vocab
    for i, sentence in enumerate(samples):
        for j, (token, tag) in enumerate(sentence[:MAX_LEN]):
            X[i, j] = vocab.get_vector(token)
            y[i,j] = tag_index[tag]
    return X, y

X_train, y_train = preprocess(train_samples)
X_val, y_val = preprocess(valid_samples)
```

-------

I replaced, `X[i, j] = vocab.get_vector(token)` with `X[i, j] = vocab._.trf_data.tensors[token]` but vocab doesn't support trf_data. Kindly advice how to proceed to get the embedding matrix with tensors.
pls note I tried with `doc._.trf_data.tensors[0].sum(axis=0)` but I am still getting issues with dimensions.

**Comment by polm** [comment]
What spaCy version are you using? Can you give the output of `spacy info`?

**Reply by urs17** [reply]
Please find the info.
============================== Info about spaCy ==============================

spaCy version    3.1.0                         
Location         c:\users\urshe\anaconda3\lib\site-packages\spacy
Platform         Windows-10-10.0.19041-SP0     
Python version   3.8.8                         
Pipelines        en_core_web_lg (3.1.0), en_core_web_sm (3.0.0), en_core_web_trf (3.1.0)

**Comment by urs17** [comment]
Hello,

Can anyone help me in this regard? I am nearing the deadline.

Thanks,

**Reply by polm** [reply]
Sorry to hear about your deadline, but this is a free support forum for free open source software. We want to help everyone but it's on a best-effort basis. Also you asked your question on a Saturday and followed up less than two days later...

You might want to look at the threads linked in #7643, or at [this blog post](https://applied-language-technology.readthedocs.io/en/latest/notebooks/part_iii/04_embeddings_continued.html). 

I'm a little confused about why you're trying to extract vectors this way though, you shouldn't normally have to do this. It sounds like you're just using spaCy to extract vectors from a Transformer and get a token alignment, but doing the actual token labeling using some non-spaCy model? You can do that but it's doing things the hard way. 

Have you tried just using the training process outlined in the docs or the [example projects](https://github.com/explosion/projects/tree/v3/tutorials)?

**Reply by urs17** [reply]
Thanks for your response and support! Yeah, I totally agree about the timeline but I am really stuck with my submission. My project is about extracting word embeddings from a transformer model (I am using roberta) and building Bi-LSTM models using with word embedding layer. I prefer using spaCy as other parts of my code developed using spaCy. I tried with the methods in the blog you mentioned but I am getting issues when I pass to the embedding layer in Keras. I will check the example projects as well. Thanks again!
